com.unity.netcode.gameobjects/Prototyping/NetworkTransform.cs

```diff
@@ -356,18 +502,146 @@ private void OnNetworkStateChanged(NetworkState oldState, NetworkState newState)
                 return;
             }
 
-            ApplyNetworkState(newState);
+            if (IsServer)
+            {
+                return; // todo use authority
+            }
+
+            var sentTime = new NetworkTime(NetworkManager.Singleton.ServerTime.TickRate, newState.SentTime);
+
+            if (newState.HasPositionX)
+            {
+                PositionXInterpolator.AddMeasurement(newState.PositionX, sentTime);
+            }
+
+            if (newState.HasPositionY)
+            {
+                PositionYInterpolator.AddMeasurement(newState.PositionY, sentTime);
+            }
+
+            if (newState.HasPositionZ)
+            {
+                PositionZInterpolator.AddMeasurement(newState.PositionZ, sentTime);
+            }
+
+            RotationInterpolator.AddMeasurement(Quaternion.Euler(newState.Rotation), sentTime);
+
+            if (newState.HasScaleX)
+            {
+                ScaleXInterpolator.AddMeasurement(newState.ScaleX, sentTime);
+            }
+
+            if (newState.HasScaleY)
+            {
+                ScaleYInterpolator.AddMeasurement(newState.ScaleY, sentTime);
+            }
+
+            if (newState.HasScaleZ)
+            {
+                ScaleZInterpolator.AddMeasurement(newState.ScaleZ, sentTime);
+            }
+
+            if (NetworkManager.Singleton.LogLevel == LogLevel.Developer)
+            {
+                var pos = new Vector3(newState.PositionX, newState.PositionY, newState.PositionZ);
+                Debug.DrawLine(pos, pos + Vector3.up + Vector3.left * Random.Range(0.5f, 2f), Color.green, k_DebugDrawLineTime, false);
+            }
         }
 
         private void Awake()
         {
             m_Transform = transform;
+
+            if (m_AllFloatInterpolators.Count == 0)
+            {
+                InitializeInterpolator<BufferedLinearInterpolatorFloat, BufferedLinearInterpolatorQuaternion, BufferedLinearInterpolatorFloatForScale>();
+            }
+
+            foreach (var interpolator in m_AllFloatInterpolators)
+            {
+                interpolator.Awake();
+                interpolator.UseFixedUpdate = UseFixedUpdate;
+            }
+
+            RotationInterpolator.Awake();
+            RotationInterpolator.UseFixedUpdate = UseFixedUpdate;
+
+            // ReplNetworkState.NetworkVariableChannel = NetworkChannel.PositionUpdate; // todo figure this out, talk with Matt/Fatih, this should be unreliable
```

Comment:
I'll defer to Matt, I haven't touched this except turning it into `NetworkDelivery` instead of `NetworkChannel`.

Reply:
@mattwalsh-unity